sklearn/linear_model/_logistic.py

```diff
@@ -865,6 +936,36 @@ def _logistic_regression_path(X, y, pos_class=None, Cs=10, fit_intercept=True,
         w0 = np.zeros((classes.size, n_features + int(fit_intercept)),
                       order='F', dtype=X.dtype)
 
+    # preconditioning for lbfgs
+    # Subtract mean, divide by standard deviation but keep scaling and
+    # mean to allow solving the original problem.
+    # The scaling is required in the gradient computation for the penalty
+    # Both scaling and mean are used later used to transform
+    # optimization results back to the original space.
+    # In the sparse case, the mean can not be subtracted and the
+    # correction is carried along as X_offset.
+    X_pre = X
+    X_scale = None
+    X_offset = None
+    if precondition and solver == 'lbfgs':
+        # FIXME this duplicates some code from _preprocess_data
+        # and should be refactored
+        X_mean, X_scale = _weighted_mean_std(X, sample_weight)
+        if sparse.issparse(X):
+            X_scale[X_scale == 0] = 1
+            if fit_intercept:
+                X_offset = -X_mean/X_scale
+            # FIXME old scipy requires conversion to sparse matrix
+            # before calling multiply
+            X_pre = X_pre.multiply(sparse.csr_matrix(1 / X_scale))
```

Comment:
I also +1 for avoiding `np.multiply`. Also doesn't inplace multiplication work with sparse (and old scipy versions)?
```suggestion
            X_pre *= sparse.csr_matrix(1 / X_scale)
```

Reply:
Good point about inplace operations.

> I prefer to keep .multiply and switch to inplace @ (that is @=) 

Aww, right nevermind those are indeed `np.matrix`, so `*` doesn't work for element wise multiplication. In other places column wise multiplication this is typically done with a dot product with a diagonal matrix, but this should also work...

Reply:
I wasn't sure if we want the dot product here or doing this of if there's an advantage to either. Quick test shows that the dia matrix is 1.5x to 2x faster for various sizes.

```python
data = np.random.normal(size=(10000,))
i = np.random.randint(0, 100003, size=(10000,))
j = np.random.randint(0, 1111, size=(10000,))
X = sp.csr_matrix((data, (i, j)))
col = np.random.normal(size=X.shape[1])
%timeit res1 = X.multiply(sp.csr_matrix(col))
%timeit res2 = X * sp.dia_matrix((col, [0]), shape=(col.shape[0], col.shape[0]))
```

Should I change to the dia matrix? I find it slightly less readable. Though we could also add a helper lol.